plutus-core/testlib/PlutusCore/Test.hs

```diff
@@ -423,47 +425,145 @@ brokenRename ::
+  whether the pass relies on global uniqueness itself, because the scoping tests rely on it anyway.
+  If the pass renames its input, and only in this case, provide 'PrerenameNo' for the 'Prerename'
+  argument, this will allow the scoping tests to ensure that the pass does indeed rename its input
+- due to a very specific design of the scoping tests some passes don't give false positives, but
+  don't get tested properly either. For example dead code elimination is a no-op within the scoping
+  tests, because internally all types/terms/programs only contain bindings that get referenced
+
+There's also 'test_scopingSpoilRenamer', this one is used to test that the scoping tests do catch
+various kinds of bugs. Don't use it with any passes that aren't renamers.
+
+All in all, in order to use the scoping tests you have to understand how they work, which is an
+unfortunate but seemingly inevitable requirement. On the bright side, it's worth the effort, because
+the tests do catch bugs occasionally.
+
+Whenever you use 'test_scopingBad' make sure to explain why, so that it's clear whether there's
+something wrong with the pass or it's just a limitation of the scoping tests.
+-}
+
+
+-- | Determines whether to perform renaming before running the scoping tests. Needed for passes that
+-- don't perform renaming themselves.
+data Prerename =
+  PrerenameYes |
+  PrerenameNo
+
+runPrerename :: TPLC.Rename a => Prerename -> a -> a
+runPrerename PrerenameYes = TPLC.runQuote . TPLC.rename
+runPrerename PrerenameNo  = id
+
 -- | Test scoping for a renamer.
 prop_scopingFor ::
-  (Pretty (t ann), Scoping t) =>
+  (PrettyPlc (t NameAnn), TPLC.Rename (t NameAnn), Scoping t) =>
+  -- | A generator of types\/terms\/programs.
   AstGen (t ann) ->
+  -- | Whether binding removal is expected for the pass.
+  BindingRemoval ->
+  -- | Whether renaming is required before running the scoping tests. Note that the scoping tests
+  -- rely on global uniqueness themselves, hence for any pass that doesn't perform renaming
+  -- internally this needs to be 'PrerenameYes'.
+  Prerename ->
+  -- | The runner of the pass.
   (t NameAnn -> TPLC.Quote (t NameAnn)) ->
   Property
-prop_scopingFor gen ren = property $ do
-  prog <- forAllPretty $ runAstGen gen
+prop_scopingFor gen bindRem preren run = withTests 1000 . property $ do
+  prog <- forAllNoShow $ runAstGen gen
   let catchEverything = unsafePerformIO . try @SomeException . evaluate
-  case catchEverything $ checkRespectsScoping (TPLC.runQuote . ren) prog of
-    Left exc         -> fail $ show exc
-    Right (Left err) -> fail $ show err
-    Right (Right ()) -> success
-
--- | Test that a good renamer does not destroy scoping.
+      prep = runPrerename preren
+  case catchEverything $ checkRespectsScoping bindRem prep (TPLC.runQuote . run) prog of
+    Left exc -> fail $ show exc
+    Right (before, after, Left err) -> fail $ concat
```

Comment:
Could return a proper error with a pretty instance instead of a tuple

Reply:
> Could return a proper error with a pretty instance instead of a tuple

Done.